src/Controller/Controller.php

```diff
@@ -544,6 +552,57 @@ public function invokeAction(Closure $action, array $args): void
         }
     }
 
+    /**
+     * Register middleware for the controller.
+     *
+     * @param string|\Closure|\Psr\Http\Server\MiddlewareInterface $middleware Middleware.
+     * @param array $options Valid options:
+     *  - `only`: (array|string) Only run the middleware for specified actions.
+     *  - `except`: (array|string) Run the middleware for all actions except the specified ones.
+     * @return void
+     * @psalm-var array{?only: string|array, ?except: string|array}
+     */
+    public function middleware($middleware, array $options = [])
```

Comment:
Should this method raise an exception if it is called after middleware have already been applied? It would be frustrating to use this method in a `beforeRender` and then think that Cake is 'broken' when you're using it wrong.

Reply:
Could be start with just documenting the fact that middlewares must be set from `initialize()` only?

I don't think we would be able to check if the middleware queue was already created without adding a new property.

Reply:
Yeah we would need another property to track whether or not middleware could be added.